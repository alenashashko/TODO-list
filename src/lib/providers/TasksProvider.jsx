import React, {useContext, useState, useCallback, useMemo} from "react";
import {nanoid} from "nanoid";

const TasksContext = React.createContext({
  tasks: []
});

export const useTasksInfo = () => {
  const value = useContext(TasksContext);

  if (value === undefined) {
    throw new Error("you should use useTasksInfo inside of TasksProvider");
  }

  return value;
};

export const TasksProvider = (props) => {
  const {children} = props;

  const [tasks, setTasks] = useState([]);

  const createTask = useCallback((text) => {
    setTasks([...tasks, {
      id: nanoid(),
      date: Date.now(),
      text
    }])
  }, [tasks]);

  const changeTask = useCallback((taskId, newText) => {
    setTasks(tasks.map((task) => {
      return task.id !== taskId
        ? task
        : {
          ...task,
          date: Date.now(),
          text: newText
        }
    }))
  }, [tasks]);

  const deleteTask = useCallback((taskId) => {
    setTasks(tasks.filter((task) => task.id !== taskId));
  }, [tasks]);

  const value = useMemo(() => ({
    tasks,
    createTask,
    changeTask,
    deleteTask
  }), [tasks, createTask, changeTask, deleteTask]);

  return <TasksContext.Provider value={value}>{children}</TasksContext.Provider>;
};
