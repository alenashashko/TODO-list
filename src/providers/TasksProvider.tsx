import {FC, createContext, useContext, useState, useCallback, useMemo} from "react";
import {nanoid} from "nanoid";
import { Task } from "components/Task/Task";

const TasksContext = createContext<{
  tasks: Task[];
  createTask: (text: string) => void;
  changeTask: (taskId: string, newText: string) => void;
  deleteTask: (taskId: string) => void;
}>({
  tasks: [],
  createTask: () => {},
  changeTask: () => {},
  deleteTask: () => {},
});

export const useTasksInfo = () => {
  const value = useContext(TasksContext);

  if (value === undefined) {
    throw new Error("you should use useTasksInfo inside of TasksProvider");
  }

  return value;
};

export const TasksProvider: FC = (props) => {
  const {children} = props;

  const [tasks, setTasks] = useState<Task[]>([]);

  const createTask = useCallback((text: string) => {
    setTasks([...tasks, {
      id: nanoid(),
      date: Date.now(),
      text
    }])
  }, [tasks]);

  const changeTask = useCallback((taskId: string, newText: string) => {
    setTasks(tasks.map((task) => {
      return task.id !== taskId
        ? task
        : {
          ...task,
          date: Date.now(),
          text: newText
        }
    }))
  }, [tasks]);

  const deleteTask = useCallback((taskId: string) => {
    setTasks(tasks.filter((task) => task.id !== taskId));
  }, [tasks]);

  const value = useMemo(() => ({
    tasks,
    createTask,
    changeTask,
    deleteTask
  }), [tasks, createTask, changeTask, deleteTask]);

  return <TasksContext.Provider value={value}>{children}</TasksContext.Provider>;
};
