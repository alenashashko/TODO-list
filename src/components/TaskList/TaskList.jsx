import React from "react";
import {Grid} from "@material-ui/core";

import {useTasksInfo} from "../../lib/providers/TasksProvider";
import {Task} from "../Task/Task";

export const TaskList = () => {
  const {tasks} = useTasksInfo();

  return (
    <Grid container spacing={3}>
      {tasks.map((task) =>
        <Grid key={task.id} item xs={12} sm={6}>
          <Task task={task} />
        </Grid>
      )}
    </Grid>
  );
};
