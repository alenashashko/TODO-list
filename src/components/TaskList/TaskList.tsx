import {FC} from "react";
import {Grid} from "@material-ui/core";

import {useTasksInfo} from "../../providers/TasksProvider";
import {Task} from "../Task/Task";

export const TaskList: FC = () => {
  const {tasks} = useTasksInfo();

  tasks.sort((a, b) => b.date - a.date);

  return (
    <Grid container spacing={3}>
      {tasks.sort().map((task:Task) =>
        <Grid key={task.id} item xs={12} sm={6}>
          <Task task={task} />
        </Grid>
      )}
    </Grid>
  );
};
