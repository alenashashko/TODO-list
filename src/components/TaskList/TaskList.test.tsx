import { render, cleanup } from "@testing-library/react";

import { TaskList } from "./TaskList";

import { useTasksInfo } from "../../providers/TasksProvider";
import { tasksMock } from "../../tests/mocks/tasks";

jest.mock("../../providers/TasksProvider");

afterEach(cleanup);

it("renders task list", () => {
  const mockedUseTasksInfo = useTasksInfo as jest.MockedFunction<
    typeof useTasksInfo
  >;

  mockedUseTasksInfo.mockReturnValue({
    ...useTasksInfo(),
    tasks: tasksMock,
  });

  const { asFragment } = render(<TaskList />);

  expect(asFragment()).toMatchSnapshot();
});
