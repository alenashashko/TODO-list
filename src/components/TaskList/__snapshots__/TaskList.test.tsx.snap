// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders task list 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6"
    >
      <div
        class="MuiPaper-root MuiCard-root makeStyles-task-1 MuiPaper-elevation1 MuiPaper-rounded"
      >
        <div
          class="MuiCardContent-root"
        >
          <p
            class="MuiTypography-root makeStyles-text-2 MuiTypography-body1"
          >
            my second task
          </p>
        </div>
        <div
          class="MuiCardActions-root makeStyles-footer-3 MuiCardActions-spacing"
        >
          <div
            class="makeStyles-footerContainer-4"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
            >
              03/09/2021 3:50 PM
            </p>
            <div
              class="makeStyles-buttons-5"
            >
              <button
                aria-label="edit task"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeSmall"
                data-testid="edit"
                tabindex="0"
                title="Edit"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
              <button
                aria-label="delete task"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorSecondary MuiIconButton-sizeSmall"
                data-testid="delete"
                tabindex="0"
                title="Delete"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6"
    >
      <div
        class="MuiPaper-root MuiCard-root makeStyles-task-1 MuiPaper-elevation1 MuiPaper-rounded"
      >
        <div
          class="MuiCardContent-root"
        >
          <p
            class="MuiTypography-root makeStyles-text-2 MuiTypography-body1"
          >
            my task
          </p>
        </div>
        <div
          class="MuiCardActions-root makeStyles-footer-3 MuiCardActions-spacing"
        >
          <div
            class="makeStyles-footerContainer-4"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
            >
              03/09/2021 3:50 PM
            </p>
            <div
              class="makeStyles-buttons-5"
            >
              <button
                aria-label="edit task"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeSmall"
                data-testid="edit"
                tabindex="0"
                title="Edit"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
              <button
                aria-label="delete task"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorSecondary MuiIconButton-sizeSmall"
                data-testid="delete"
                tabindex="0"
                title="Delete"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
